### PR TITLE
[ApplyNSICategoriesPipeline] filter_cc to consider category applied in spider

### PR DESCRIPTION
**Pull request by panasonic-mafia**
A small structural fix for some cases from #6141:

- `enterprise` spider used to apply multiple categories because `filter_cc` used to prioritise country specific matches even if applied category is different.
- `indigo` - failed to match NSI in France because there were multiple France-specific records in another categories than specified in spider.
- `magnit_ru` - used to prefer Russia-specific `Magnit Cosmetics` NSI record for supermarkets.

I would like to ask for review from the folks @Cj-Malone @davidhicks please.

### CODE REVIEW

File: locations/pipelines/apply_nsi_categories.py [hunk 1]
```diff
@@ -68,9 +68,14 @@ def filter_cc(self, matches, cc) -> []:
...
 
+        if categories:
+            # Find better match if category is supplied
+            includes = self.filter_categories(includes, categories)
+            globals_matches = self.filter_categories(globals_matches, categories)
+
         return includes or globals_matches
 
-    def filter_categories(self, matches, tags) -> []:
+    def filter_categories(self, matches, tags: dict) -> list:
```
[the remark this answers is not in the document]
Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py:78]: I suggest too that this function should perform some fuzzy matching for categories such that if a match has `{"amenity": "cafe;cinema"}` then it'd get returned if `tags={"amenity": "cafe"}`.

**Review by davidhicks — CHANGES_REQUESTED** [review]
Looks good overall. I've suggested some changes that are more enhancements to existing code to make it more accurate, especially due to apply_nsi_categories now needing a more accurate category matching approach.

File: locations/pipelines/apply_nsi_categories.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ def process_item(self, item, spider):
         spider.crawler.stats.inc_value("atp/nsi/match_failed")
         return item
 
-    def filter_cc(self, matches, cc) -> []:
+    def filter_cc(self, matches, cc, categories: dict = None) -> list:
```
[the remark this answers is not in the document]
Comment (Cj-Malone) [reply locations/pipelines/apply_nsi_categories.py:53]: It really can't. Nothing in ATP supports `{"amenity": "cafe;cinema"}` other than your addition to `apply_category` and it causing headaches.

File: locations/pipelines/apply_nsi_categories.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cj-Malone) [reply locations/pipelines/apply_nsi_categories.py:78]: https://wiki.openstreetmap.org/wiki/Semi-colon_value_separator#When_NOT_to_use

Order is questionable: https://github.com/osmlab/name-suggestion-index/issues/8397 https://wiki.openstreetmap.org/wiki/Talk:Key:cuisine#Significance_of_value_order

File: locations/pipelines/apply_nsi_categories.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py:53]: Thanks for the link to https://wiki.openstreetmap.org/wiki/Semi-colon_value_separator#When_NOT_to_use

I'd be happy if we then fix `apply_category` to only allow one value, and overwrite the value if `apply_category` is called twice.

[the remark this answers is not in the document]
Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py:53]: @Cj-Malone I get 1314 records in NSI's dist/nsi.json that contain more than one top-level tag as defined in ATP's categories.py:

`jq '[(..|.tags?)|select(([.amenity, .club, .craft, .dark_store, .emergency, .healthcare, .highway, .landuse, .leisure, .man_made, .office, .public_transport, .shop, .tourism, .aeroway, .railway, .waterway, .telecom|select(.)]|length) > 1)]|length' nsi.json`

Command to output all 1314 matches:
`jq '(..|.tags?)|select(([.amenity, .club, .craft, .dark_store, .emergency, .healthcare, .highway, .landuse, .leisure, .man_made, .office, .public_transport, .shop, .tourism, .aeroway, .railway, .waterway, .telecom|select(.)]|length) > 1)' nsi.json`

**Comment by panasonic-mafia** [comment]
@davidhicks Thanks for feedback. I added better type annotations for all methods I touched in this PR. Multiple top-level tags and semicolon-delimited categories are a bit out of scope of this PR. But I'm happy to take a look in multiple top-level tags at some point.

File: locations/pipelines/apply_nsi_categories.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Cj-Malone) [reply locations/pipelines/apply_nsi_categories.py]: Yeah, so there is a bunch of OSM details here. `tourism=yes` isn't a top level tag, but added to suggest another category is of importance to tourists.

`public_transport` largely duplicates `highway=bus_stop`, `railway=*`... See Public Transport Version 1, Public Transport Version 2, Public Transport Version 2.5, Public Transport Version 2 new for more drama if you want.
`healthcare` is another case where large deprecations where attempted, only for some people to object and then everyone get bored half way through. They will still be in the main tree eg `amenity/hospital`, but also have `healthcare=hospital` on them.
`club` being a top level tag is debatable, there is even a argument about if they should be mapped in OSM as they aren't really physical. Something like `amenity=community_centre` is often the physical thing, and if it's exclusive the club tag could go with it, if there are multiple clubs that share the centre, they would be separate with 1 `amenity=community_centre`.
`emergency` is a mess.

If you take into account the above details, that number will be much smaller, but not 0. Fedex is an example that annoys me, copyshop and post office. I think it's an issue with NSI. However it's not worth an argument.

We were actually warned about using the OSM categories like we are. But we didn't listen and now we have Schrodingers categories. :( https://github.com/alltheplaces/alltheplaces/issues/3585#issuecomment-1104486807

[the remark this answers is not in the document]
Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py]: The good news at least is ";" is never used within NSI in any of the top-level-tags defined by ATP:

`jq '(..|.tags?)|select(.amenity, .club, .craft, .dark_store, .emergency, .healthcare, .highway, .landuse, .leisure, .man_made, .office, .public_transport, .shop, .tourism, .aeroway, .railway, .waterway, .telecom|select(.)|select(contains(";")))' nsi.json`

**Comment by panasonic-mafia** [comment]
Ok, now it's good to be merged.

File: locations/pipelines/apply_nsi_categories.py [hunk 3]
```diff
@@ -50,7 +50,13 @@ def process_item(self, item, spider):
         spider.crawler.stats.inc_value("atp/nsi/match_failed")
         return item
 
-    def filter_cc(self, matches, cc) -> []:
+    def filter_cc(self, matches: list[dict], cc: str, categories: dict = None) -> list[dict]:
+        """Filter matches by country code, attempt to find a better match if category is supplied.
+        :param matches: list of matches from NSI
+        :param cc: country code in lower case
+        :param categories: category tags
```
Comment (davidhicks) [inline locations/pipelines/apply_nsi_categories.py:57]: Can the documentation for `categories` clarify that categories have to be an exact match (e.g. if two tags are supplied, both tags have to present on the NSI item for a match to occur)?

File: locations/pipelines/apply_nsi_categories.py [hunk 4]
```diff
@@ -68,9 +74,19 @@ def filter_cc(self, matches, cc) -> []:
...
+            includes = self.filter_categories(includes, categories)
+            globals_matches = self.filter_categories(globals_matches, categories)
+
         return includes or globals_matches
 
-    def filter_categories(self, matches, tags) -> []:
+    def filter_categories(self, matches: list[dict], tags: dict) -> list[dict]:
+        """Filter matches by category tags.
+        :param matches: list of matches from NSI
+        :param tags: category tags
```
Comment (davidhicks) [inline locations/pipelines/apply_nsi_categories.py:87]: Can the documentation for `tags` clarify that categories have to be an exact match (e.g. if two tags are supplied, both tags have to present on the NSI item for a match to occur)?

**Review by davidhicks — CHANGES_REQUESTED** [review]

File: locations/pipelines/apply_nsi_categories.py (hunk 3 above)
Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py:57]: (this strict matching is what the test cases are testing for in this proposed change)

Comment (davidhicks) [reply locations/pipelines/apply_nsi_categories.py:57]: Suggest too that documentation also clarifies that the strict matching is only done for `top_level_tags` from categories.py and any other tags are ignored in the matching.

Comment (panasonic-mafia) [reply locations/pipelines/apply_nsi_categories.py:57]: Yep. This doc update makes sense in `filter_categories`.

Comment (panasonic-mafia) [reply locations/pipelines/apply_nsi_categories.py:57]: > Suggest too that documentation also clarifies that the strict matching is only done for top_level_tags from categories.py and any other tags are ignored in the matching.

Name `get_category_tags` is already tells us that only category tags are used.

**Review by davidhicks — APPROVED** [review]

**Comment by panasonic-mafia** [comment]
@iandees may kindly I ask to merge this one please? It was approved :)

**Pull request merged**